app/kuma-dp/pkg/dataplane/metrics/server.go

```diff
@@ -153,44 +174,142 @@ func rewriteMetricsURL(address string, port uint32, path string, queryModifier Q
 func (s *Hijacker) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
 	ctx := req.Context()
 	out := make(chan []byte, len(s.applicationsToScrape))
+	contentTypes := make(chan expfmt.Format, len(s.applicationsToScrape))
 	var wg sync.WaitGroup
 	done := make(chan []byte)
 	wg.Add(len(s.applicationsToScrape))
 	go func() {
 		wg.Wait()
-		close(done)
 		close(out)
+		close(contentTypes)
+		close(done)
 	}()
 
 	for _, app := range s.applicationsToScrape {
 		go func(app ApplicationToScrape) {
 			defer wg.Done()
-			out <- s.getStats(ctx, req, app)
+			content, contentType := s.getStats(ctx, req, app)
+			out <- content
+
+			// It's possible to track the highest priority content type seen,
+			// but that would require mutex.
+			// I would prefer to calculate it later at one go
+			contentTypes <- contentType
 		}(app)
 	}
 
 	select {
 	case <-ctx.Done():
 		return
 	case <-done:
-		// default format returned by prometheus
-		writer.Header().Set("content-type", string(expfmt.FmtText))
-		for resp := range out {
-			if _, err := writer.Write(resp); err != nil {
-				logger.Error(err, "error while writing the response")
-			}
-			if _, err := writer.Write([]byte("\n")); err != nil {
-				logger.Error(err, "error while writing the response")
-			}
+		selectedCt := selectContentType(contentTypes, req.Header)
+		writer.Header().Set(hdrContentType, string(selectedCt))
+
+		// aggregate metrics of target applications and attempt to make them
+		// compatible with FmtOpenMetrics if it is the selected content type.
+		metrics := processMetrics(out, selectedCt)
+		if _, err := writer.Write(metrics); err != nil {
+			logger.Error(err, "error while writing the response")
+		}
+	}
+}
+
+func processMetrics(metrices <-chan []byte, contentType expfmt.Format) []byte {
+	buf := new(bytes.Buffer)
+
+	for metrics := range metrices {
+		// remove the EOF marker from the metrics, because we are
+		// merging multiple metrics into one response.
+		metrics = bytes.ReplaceAll(metrics, []byte("# EOF"), []byte(""))
+
+		if _, err := buf.Write(metrics); err != nil {
+			logger.Error(err, "error while writing the response to temporary buffer")
+		}
+		if _, err := buf.Write([]byte("\n")); err != nil {
+			logger.Error(err, "error while writing the response")
 		}
 	}
+
+	str := processNewlineChars(buf.String(), true, true)
+	str = fmt.Sprintf("%s\n", str)
+
+	// if the content type is not OpenMetrics, we don't need to add EOF marker
+	if !(contentType == expfmt.FmtOpenMetrics_1_0_0 || contentType == expfmt.FmtOpenMetrics_0_0_1) {
+		return []byte(str)
+	}
+
+	// make metrics OpenMetrics compliant
+	return []byte(fmt.Sprintf("%s# EOF\n", str))
+}
+
+// processNewlineChars processes the newline characters in the text.
+// If dedup is true, it replaces multiple newline characters with a single newline character.
+// If trim is true, it trims the leading and trailing newline characters.
+func processNewlineChars(text string, dedup, trim bool) string {
```

Comment:
This is only called with `true, true` isn't it?

Reply:
yes.. I could have provided `processNewlineChars()` with out params. I just did this way to make it configurable even if at this moment we are just invoking it with `true, true`